chain/epoch-manager/src/proposals.rs

```diff
@@ -270,15 +269,51 @@ mod old_validator_selection {
 
     fn shuffle_duplicate_proposals(dup_proposals: &mut Vec<u64>, rng_seed: RngSeed) {
         let mut rng: Hc128Rng = SeedableRng::from_seed(rng_seed);
-        dup_proposals.shuffle(&mut rng);
+        for i in (1..dup_proposals.len()).rev() {
+            dup_proposals.swap(i, gen_index_old(&mut rng, (i + 1) as u64) as usize);
+        }
     }
 
+    fn gen_index_old(rng: &mut Hc128Rng, bound: u64) -> u64 {
+        // This is a simplified copy of the rand gen_index implementation to ensure that
+        // upgrades to the rand library will not cause a change in the shuffling behavior.
+        let zone = u64::MAX - (u64::MAX - bound + 1) % (bound);
+        loop {
+            let v = rng.next_u64();
+            let mul = (v as u128) * (bound as u128);
+            let (hi, lo) = ((mul >> 64) as u64, mul as u64);
+            if lo < zone {
+                return hi;
+            }
+        }
+    }
```

Comment:
If we are feeling enterpisey, I'd probably extracted this code into a separate module/type `EpochManagerRng` to make sure its public API *doesn't* expose `rand`. Right now this code has an implicit invariant of "don't use any rand methods besides from_seed and next_u64", and I'd love for that invariant to have teeth.

Reply:
Sorry not exactly sure what you mean - the module is already named old_* and these two functions are not public.f

Reply:
Ah, sorry, I missed two points:

* that this is already a separate module
* that we no longer relying on randomness here for the latest protocol

I was fearing that we still use `rand`, and need to guard aginst future code changes.